fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,541 @@
+ * @return The node argument linked list
+ */
+ASTNodeArg* parseArgList(TokenIter &tokenIt) {
+    ASTNodeArg* currentArg = NULL;
+    ASTNodeArg* argList = NULL;
+
+    while (tokenIt->getType() != Token::LPAREN && tokenIt->getType() != Token::RPAREN) {
+        if (tokenIt->getType() == Token::ID && tokenIt.peek().getType() == Token::EQUALS) {
+            char * node_args_name = new char[tokenIt->getValue().size() + 1];
+            strcpy(node_args_name, tokenIt->getValue().c_str());
+            ++tokenIt; // consume ID
+            ++tokenIt; // consume EQUALS
+            if (tokenIt->getType() == Token::LPAREN_SQ) {
+                ++tokenIt; // consume LPAREN_SQ
+                currentArg = createNodeArg(node_args_name, parseValueList(tokenIt), NULL);
+            } else {
+                currentArg = createNodeArg(node_args_name, buildNodeValue(*tokenIt), NULL);
+            }
+        } else if (isValue(tokenIt)) {
+            currentArg = createNodeArg("", buildNodeValue(*tokenIt), NULL);
+        } else {
+            throw ParserFailure("Expecting value token but got " + tokenIt->getValue());
+        }
+        if (argList == NULL) {
+            argList = currentArg;
+        } else {
+            appendSiblingArg(argList, currentArg);
+        }
+        ++tokenIt; // consume BECOMES
+    }
+    return argList;
+}
+
+/* @brief Parse and build the AST. 
+ * When the function returns, the token iterator points to the token immediately
+ * after the ')' of the parsed s-expression.
+ * Exceptions can be thrown when the parsed grammer is incorrect in the following situation:
+ *   1. LPAREN followed by end of node
+ *   2. LPAREN followed by none ID type
+ *   3. Missing RPAREN at the end of node
+ *
+ * @param tokenIt The tokens contains the AST info
+ * @return The root of the AST
+ */
+ASTNode* buildAST(TokenIter &tokenIt) {
+    ASTNode* siblingsList = NULL;
+    while (!tokenIt.isEnd() && tokenIt->getType() == Token::LPAREN) {
+        ASTNodeArg* argList = NULL;
+        ASTNode* currentchild = NULL;
+        ASTNode* childrenList = NULL;
+        ++tokenIt; // consume LPAREN
+        if (tokenIt.isEnd()) {
+            throw ParserFailure("Token stream ended unexpectedly after LPAREN");
+        } else if (tokenIt->getType() != Token::ID) {
+            throw ParserFailure("The next token of LPAREN should be IDENTIFIER, while current value is " +  tokenIt->getValue());
+        }
+        char* node_name = new char[tokenIt->getValue().size() + 1];
+        strcpy(node_name, tokenIt->getValue().c_str());
+        ++tokenIt; // consume ID
+        if (isValue(tokenIt)) {
+            argList = parseArgList(tokenIt);
+        }
+        currentchild = buildAST(tokenIt);
+        if (childrenList == NULL) {
+            childrenList = currentchild;
+        } else {
+            appendSiblingNode(childrenList, currentchild);
+        }
+        if ((!tokenIt.isEnd() && tokenIt->getType() != Token::RPAREN) || tokenIt.isEnd()) {
```

Comment:
Because of shortcutting, I think this expression can be simplified to
```suggestion
        if (tokenIt.isEnd() || tokenIt->getType() != Token::RPAREN) {
```